### PR TITLE
Малышев Владимир. Задача 2. Вариант 15. Ленточная горизонтальная схема

### PR DESCRIPTION
**Pull request by dangerulya**
### Последовательная версия

**Подготовка:**
1. Определяем размеры матрицы (строки `rows` и столбцы `cols`).
2. Определяем размер вектора `vector_size`.
3. Генерируем случайную матрицу `matrix` размером `rows x cols`.
4. Генерируем случайный вектор `vector` размером `vector_size`.
5. Создаём вектор `result` размером `rows` для хранения результата умножения.

**Итеративный процесс:**
1. Для каждой строки `i` матрицы:
   - Инициализируем элемент `result[i]` нулем.
   - Для каждого элемента `j` в строке `i`:
     - Увеличиваем `result[i]` на произведение `matrix[i][j] * vector[j]`.

**Результат:**
- Вектор `result` содержит результат умножения матрицы на вектор.

### Параллельная версия

**Подготовка:**
1. Определяем размеры матрицы (строки `rows` и столбцы `cols`).
2. Определяем размер вектора `vector_size`.
3. Генерируем случайную матрицу `matrix` размером `rows x cols`.
4. Генерируем случайный вектор `vector` размером `vector_size`.
5. Создаём вектор `result` размером `rows` для хранения результата умножения.
6. Распределяем строки матрицы между процессами. Каждый процесс получает часть строк матрицы.

**Итеративный процесс:**
1. Каждый процесс обрабатывает свои строки матрицы:
   - Для каждой строки `i` в локальной части матрицы:
     - Инициализируем элемент `local_result[i]` нулем.
     - Для каждого элемента `j` в строке `i`:
       - Увеличиваем `local_result[i]` на произведение `local_matrix[i][j] * vector[j]`.
2. Собираем результаты от всех процессов в корневой процесс.

**Результат:**
- Вектор `result` в корневом процессе содержит результат умножения матрицы на вектор.

### CODE REVIEW

File: tasks/mpi/malyshev_lent_horizontal/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+  for (auto &el : data) {
+    el = -200 + gen() % (300 + 200 + 1);
+  }
+
+  return data;
+}
+
+}  // namespace malyshev_lent_horizontal
+
+TEST(malyshev_lent_horizontal, test_vertical_stretched_matrix_100x75) {
```
Comment (myffii) [inline tasks/mpi/malyshev_lent_horizontal/func_tests/main.cpp:40]: please add tests for zero matrix\vector

**Review by myffii — CHANGES_REQUESTED** [review]

File: tasks/mpi/malyshev_lent_horizontal/func_tests/main.cpp (hunk 1 above)
Comment (dangerulya) [reply tasks/mpi/malyshev_lent_horizontal/func_tests/main.cpp]: new test added

File: tasks/seq/malyshev_lent_horizontal/func_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+  for (auto &el : data) {
+    el = -200 + gen() % (300 + 200 + 1);
+  }
+
+  return data;
+}
+
+}  // namespace malyshev_lent_horizontal
+
```
Comment (keshaproger) [inline tasks/seq/malyshev_lent_horizontal/func_tests/main.cpp:38]: please add tests for zero matrix\vector in seq

Comment (dangerulya) [reply tasks/seq/malyshev_lent_horizontal/func_tests/main.cpp]: new test added

**Review by keshaproger — APPROVED** [review]

**Review by myffii — APPROVED** [review]

File: tasks/mpi/malyshev_v_lent_horizontal/func_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+namespace malyshev_lent_horizontal {
+
+std::vector<std::vector<int32_t>> generateRandomMatrix(uint32_t rows, uint32_t cols) {
+  std::random_device dev;
+  std::mt19937 gen(dev());
+  std::vector<std::vector<int32_t>> data(rows, std::vector<int32_t>(cols));
+
+  for (auto &row : data) {
+    for (auto &el : row) {
+      el = -200 + gen() % (300 + 200 + 1);
```
Comment (allnes) [inline tasks/mpi/malyshev_v_lent_horizontal/func_tests/main.cpp:19]: please parametrize interval of random

File: tasks/mpi/malyshev_v_lent_horizontal/func_tests/main.cpp [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+  return data;
+}
+
+std::vector<int32_t> generateRandomVector(uint32_t size) {
+  std::random_device dev;
+  std::mt19937 gen(dev());
+  std::vector<int32_t> data(size);
+
+  for (auto &el : data) {
+    el = -200 + gen() % (300 + 200 + 1);
```
Comment (allnes) [inline tasks/mpi/malyshev_v_lent_horizontal/func_tests/main.cpp:32]: please parametrize it

File: tasks/mpi/malyshev_v_lent_horizontal/func_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+  std::vector<std::vector<int32_t>> randomMatrix;
+  std::vector<int32_t> randomVector;
+  std::vector<int32_t> mpiResult;
+
+  std::shared_ptr<ppc::core::TaskData> taskDataPar = std::make_shared<ppc::core::TaskData>();
+  malyshev_lent_horizontal::TestTaskParallel taskMPI(taskDataPar);
+
+  if (world.rank() == 0) {
+    randomMatrix = malyshev_lent_horizontal::generateRandomMatrix(rows, cols);
+    randomVector = malyshev_lent_horizontal::generateRandomVector(cols);
```
Comment (allnes) [inline tasks/mpi/malyshev_v_lent_horizontal/func_tests/main.cpp:294]: please check situation when vector and matrix sizes not equal (e.g. 7, 8, 9)

Comment (dangerulya) [reply tasks/mpi/malyshev_v_lent_horizontal/func_tests/main.cpp]: Done

File: tasks/mpi/malyshev_v_lent_horizontal/func_tests/main.cpp (hunk 4 above)
Comment (dangerulya) [reply tasks/mpi/malyshev_v_lent_horizontal/func_tests/main.cpp]: Done

File: tasks/mpi/malyshev_v_lent_horizontal/func_tests/main.cpp (hunk 3 above)
Comment (dangerulya) [reply tasks/mpi/malyshev_v_lent_horizontal/func_tests/main.cpp]: Done

**Review by aobolensk — APPROVED** [review]

**Review by allnes — APPROVED** [review]
